File: 2022/C++/Day-3/Part-1.cpp
#include <set>
#include <string>
#include <fstream>
#include <iostream>

int main() {
    std::ios_base::sync_with_stdio(false);
    std::ifstream inputFile("input.txt");
    //-------------------------------------
    int priority_sum(0);

    std::string line;
    while (std::getline(inputFile, line)) {
        if (line.empty()) break;
        int mid = (line.length() / 2);

        std::set prefix(line.begin(), line.begin() + mid);
        std::set suffix(line.begin() + mid, line.end());

        for (char ch : prefix) {
            if (suffix.count(ch)) {
                priority_sum += (isupper(ch)) ? (ch - 38) : (ch - 96);
                break;
            }
        }
    }

    std::cout << priority_sum << "\n";
    //-------------------------------------
    return 0;
}

File: 2022/Python/Day-3/Part-1.py
from helper_functions import get_relative_path


def calculate_priority_sum(line):
    mid = len(line) // 2
    prefix = set(line[:mid])
    suffix = set(line[mid:])
    common = list(prefix.intersection(suffix))[0]
    return (ord(common) - 38) if common.isupper() else (ord(common) - 96)



def main():
    priority_sum = 0

    with open(get_relative_path("input.txt")) as file:
        while line := file.readline().strip():
                priority_sum += calculate_priority_sum(line)
                
    print(priority_sum)


if __name__ == "__main__":
    main()
